src/provider/pedersen.rs

```diff
@@ -33,10 +34,40 @@ impl<G: Group> Clone for CommitmentKey<G> {
   }
 }
 
-impl<G: Group> Len for CommitmentKey<G> {
+impl<G: Group> CommitmentKeyTrait for CommitmentKey<G> {
   fn length(&self) -> usize {
     self.ck.len()
   }
+
+  fn is_prefix_of(&self, other: &Self) -> bool {
+    other.ck.starts_with(&self.ck)
+  }
+
+  fn encode<W: std::io::Write>(&self, write: &mut W) -> io::Result<()> {
+    // TODO: unwrap()
+    unsafe { abomonation::encode(self, write) }
+  }
+
+  fn decode(bytes: &mut [u8], n: usize) -> io::Result<Self> {
+    // Note: this is unsafe but stable; it only assumes that `bytes` is any valid commitment key
+    unsafe {
+      // Ideally, we would like to allocate only one commitment key, because if we simply loaded
+      // all the bytes from the reader, we might be tapping into a cached commitment key with
+      // >=10^7 elements and kill the process. Instead, we need to manipulate the bytes to convince
+      // `abomonation::decode` to only decode what we need by manipulating the `Vec` header.
+      let mut test_ck = Self { ck: Vec::new() };
+      test_ck.ck.set_len(n);
+      let header_size = mem::size_of::<Self>();
+      let header: &[u8] =
+        std::slice::from_raw_parts(&test_ck as *const _ as *const u8, header_size);
+      bytes[..header_size].copy_from_slice(header);
```

Comment:
I suspect this can fit within a `MaybeUninit` usage: https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#initializing-an-array-element-by-element